### PR TITLE
Esm2 on Sagemaker Hyperpod

### PR DESCRIPTION
**Pull request by awsankur**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by KeitaW** [comment]
Do we have any SMHP specific feature in this test case?
If not we may organize the test case per scheduler:
```
23.esm
├── kubernetes
└── slurm
```
https://github.com/aws-samples/awsome-distributed-training/pull/381

File: 3.test_cases/23.SMHP-esm2/README.md [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+```
+
+### 5.1 Accelerate training with torch.compile
+
+HuggingFace provides an easy to use [Trainer](https://huggingface.co/docs/transformers/en/main_classes/trainer) class that also provides an option to compile the model graph. For more details on torch.compile, follow this [blog](https://pytorch.org/blog/maximizing-training-throughput/) We notice a speedup of 43% when pre-training ESM2 with torch.compile:
+
+|  Model | device_batch_size | num_nodes | torch.compile |     Instance   |   Throughput   |
+|:------:|:-----------------:|:---------:|:-------------:| :------------: | :------------: |
+|  ESM2  |         8         |     2     |       No      |  g5.12xlarge   |  160 samples/s |
```
Comment (KeitaW) [inline 3.test_cases/23.SMHP-esm2/README.md:147]: The set up instruction advise to use 24xl but actually 12xl was used?

File: 3.test_cases/23.SMHP-esm2/README.md [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+
+## What is ESM-2?
+[ESM-2](https://www.biorxiv.org/content/10.1101/2022.07.20.500902v1) is a pLM trained using unsupervied masked language modelling on 250 Million protein sequences by researchers at [Facebook AI Research (FAIR)](https://www.biorxiv.org/content/10.1101/2022.07.20.500902v1). It is available in several sizes, ranging from 8 Million to 15 Billion parameters. The smaller models are suitable for various sequence and token classification tasks. The FAIR team also adapted the 3 Billion parameter version into the ESMFold protein structure prediction algorithm. They have since used ESMFold to predict the struture of [more than 700 Million metagenomic proteins](https://esmatlas.com/about).
+
+ESM-2 is a powerful pLM. We will demonstrate how to use QLoRA to fine-tune ESM-2 on g5.24xlarge instances. We will use ESM-2 to predict [subcellular localization](https://academic.oup.com/nar/article/50/W1/W228/6576357?login=false). Understanding where proteins appear in cells can help us understand their role in disease and find new drug targets.
```
Comment (KeitaW) [inline 3.test_cases/23.SMHP-esm2/README.md:10]: Is this test case demonstrating pretraining? or finetuning? I believe latter but the title states former.

File: 3.test_cases/23.SMHP-esm2/README.md [hunk 3]
```diff
@@ -0,0 +1,168 @@
+# How to pretrain ESM2 with SageMaker Hyperpod using Amazon G5 instances
```
Comment (KeitaW) [inline 3.test_cases/23.SMHP-esm2/README.md:1]:
```suggestion
# How to finetune ESM2 with SageMaker Hyperpod using Amazon G5 instances
```

File: 3.test_cases/23.SMHP-esm2/3.train_fsdp.sh [hunk 4]
```diff
@@ -0,0 +1,72 @@
+#!/bin/bash
+
+#SBATCH --job-name=esm2-accelerate
+#SBATCH -D .
```
Comment (KeitaW) [inline 3.test_cases/23.SMHP-esm2/3.train_fsdp.sh:4]:
```suggestion
```
This line may not be necessary.
> -D, --chdir=<directory>
> Set the working directory of the batch script to directory before it is executed. The path can be specified as full path or relative path to the directory where the command is executed.

File: 3.test_cases/23.SMHP-esm2/2.train_ddp.sh [hunk 5]
```diff
@@ -0,0 +1,81 @@
+#!/bin/bash
+
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: MIT-0
+
+#SBATCH --nodes=2 # number of nodes to use
+#SBATCH --job-name=FSDP # name of your job
```
Comment (KeitaW) [inline 3.test_cases/23.SMHP-esm2/2.train_ddp.sh:7]:
```suggestion
#SBATCH --job-name=DDP # name of your job
```

File: 3.test_cases/23.SMHP-esm2/3.train_fsdp.sh [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+#SBATCH --output=accelerate-%x.%j.out
+#SBATCH --nodes=2              # number of nodes
+#SBATCH --ntasks-per-node=1         # number of MP tasks
```
Comment (KeitaW) [inline 3.test_cases/23.SMHP-esm2/3.train_fsdp.sh:7]: Maybe you want 
```suggestion
#SBATCH --exclusive # job has exclusive use of the resource, no sharing
```
instead?

**Comment by perifaws** [comment]
@awsankur @KeitaW are we good on this?
